### PR TITLE
Story CORE-418 | Create Sidecar Factory

### PR DESCRIPTION
**Pull request by LucasStarlingdePaulaSalles**
# Description

Kind: story

Section: memory brokers

Summary: Created a abstract sidecar factory and configured it to work with broker manager.

## Changelog

-    features:
    - developed a function type to represent sidecar factories
    - declared a interface for sidecars and implemented its structure as an abstract factory
    - developed a Manager interface to envelop BrokerManager and  AbstractBrokerFactory
    - implemented Manager's struct
    - implemented SidecarInterface's methods
    - tested previously mentioned methods
- fixes:
    - moved broker management interfaces and structures to a more appropriate directory

### CODE REVIEW

File: cmd/insprd/memory/fake/brk_manager.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+			fail:   failErr,
+			broker: &metabroker.Brokers{},
+		},
+	}
+}
+
+// Brokers mock of broker interface
+func (bm *BrkManager) Brokers() brokers.BrokerInterface {
+	return &bm.brokers
+}
```
Comment (franciscobonand) [inline cmd/insprd/memory/fake/brk_manager.go]: Can't this file be named "broker_manager.go"?

File: cmd/insprd/memory/brokers/factory.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	var ret []k8s.DeploymentOption
+	for _, broker := range getAllSidecarNames(app) {
+		if option, ok := f[broker]; ok {
+			ret = append(ret, option(app, getAvailiblePorts()))
+		} else {
+			panic("broker not allowed")
+		}
+	}
+	return ret
+}
```
Comment (franciscobonand) [inline cmd/insprd/memory/brokers/factory.go]: Might be interesting to test this function

File: cmd/insprd/memory/brokers/manager.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+var brokerMemory Manager
+
+// GetBrokerMemory allows for connection with BrokersManager sigleton
+func GetBrokerMemory() Manager {
+	if brokerMemory == nil {
+		brokerMemory = &BrokerManager{}
+	}
+	return brokerMemory
+}
```
Comment (franciscobonand) [inline cmd/insprd/memory/brokers/manager.go]: Test this as well

**Review by franciscobonand — CHANGES_REQUESTED** [review]

File: cmd/insprd/memory/brokers/factory.go (hunk 2 above)
Comment (LucasStarlingdePaulaSalles) [reply cmd/insprd/memory/brokers/factory.go]: This is just a placeholder, it will acctually be implemented after channel changes

**Review by franciscobonand — APPROVED** [review]
LGTM

File: cmd/insprd/memory/brokers/factory.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import (
+	"github.com/inspr/inspr/pkg/ierrors"
+	"github.com/inspr/inspr/pkg/meta"
+	"github.com/inspr/inspr/pkg/operator/k8s"
+	"github.com/inspr/inspr/pkg/sidecar/models"
+)
+
+type brokerFactory map[string]models.SidecarFactory
+
+var f brokerFactory
```
Comment (luiz-couto) [inline cmd/insprd/memory/brokers/factory.go]: Change this to a more descriptive name

**Review by luiz-couto — CHANGES_REQUESTED** [review]

File: cmd/insprd/memory/brokers/factory.go (hunk 4 above)
Comment (LucasStarlingdePaulaSalles) [reply cmd/insprd/memory/brokers/factory.go]: Thanks for correcting my dumb ways

**Review by luiz-couto — APPROVED** [review]

**Review by nicholaspcr — APPROVED** [review]

**Pull request merged**